### PR TITLE
Update: [AEA-4180] - Update custom PSU transformation lambda to handle DispenseComplete

### PR DESCRIPTION
**Pull request by wildjames**
## Summary

- Routine Change

### Details

Two additions.

#### Update the schema for prescription items

The item entry in an incoming cPSU request can contain a `completedStatus` field, which can take the values `["Cancelled", "Expired", "NotDispensed", "Collected"]`. The schema has been updated to include this option (it is not mandatory, and may not be given).

#### Handle excess messages

External providers will send additional messages after we consider a prescription to be completed. We may receive updates to the cPSU translation layer that mark an item as both `status: "DispensingComplete"`, and `completedStatus: "<Cancelled, Expired, NotDispensed>"`. In these cases, we will already have received a previous update flagging the item as failed in some way, so we want to ignore such messages. To handle that, we simply do not populate the entry array with the relevant item. 

Note that downstream interfaces need to behave well with empty `entry` arrays, for the case where all items are ignored.

### CODE REVIEW

File: packages/cpsuLambda/src/schema/format_1/transformer.ts [hunk 1]
```diff
@@ -10,13 +10,16 @@ import {v4 as uuidv4} from "uuid"
 import {Transformer} from "../../handler"
 import {wrap_with_status} from "../../utils"
 import {Md5} from "ts-md5"
+import {Logger} from "@aws-lambda-powertools/logger"
+
+const logger = new Logger({serviceName: "CustomUpdatePrescriptionStatus"})
```
Comment (originalphil) [inline packages/cpsuLambda/src/schema/format_1/transformer.ts:15]: We should use the _logger passed into the transformer function instead of creating this new one. There's an append_headers function that makes sure it has all the required headers for tracking and it'll have the correct LOG_LEVEL set from the lambda env vars.

File: packages/cpsuLambda/src/schema/format_1/transformer.ts [hunk 2]
```diff
@@ -79,9 +82,20 @@ export function populateTemplate(
   template: string,
   prescriptionItem: itemType,
   prescriptionDetails: requestType
-): Result<BundleEntry<Task>, string> {
+): Result<BundleEntry<Task>, string> | undefined {
   const entry = JSON.parse(template) as BundleEntry<Task>
 
+  if (prescriptionItem.status === "DispensingComplete") {
+    const forbiddenStatuses = ["NotDispensed", "Expired", "Cancelled"]
```
Comment (originalphil) [inline packages/cpsuLambda/src/schema/format_1/transformer.ts:89]: Does this need 'Collected' adding?

File: packages/cpsuLambda/src/schema/format_1/transformer.ts [hunk 3]
```diff
@@ -79,9 +82,20 @@ export function populateTemplate(
...
+
+    if (prescriptionItem.completedStatus && forbiddenStatuses.includes(prescriptionItem.completedStatus)) {
+      logger.info("Skipping data store update for DispensingComplete with NotDispensed or Expired status", {
```
Comment (originalphil) [inline packages/cpsuLambda/src/schema/format_1/transformer.ts:92]: Does this need the other two forbidden statuses adding? Maybe reword it to clarify that it won't send that update on, to the PSU lambda, rather than mention the data store.

File: packages/cpsuLambda/tests/format_1/business_states.test.ts [hunk 4]
```diff
@@ -44,52 +48,96 @@ describe("populateTemplate function", () => {
...
     },
     {
       itemStatus: "DispensingComplete",
       deliveryType: "Robot Collection",
       expectedBusinessStatus: "Dispatched",
-      expectedTaskStatus: "completed"
+      expectedTaskStatus: "completed",
+      expectItemDefined: true
+    },
+    {
```
Comment (originalphil) [inline packages/cpsuLambda/tests/format_1/business_states.test.ts:61]: Additions in here are good. Can one be added to testHandler too?

File: packages/cpsuLambda/src/schema/format_1/transformer.ts (hunk 1 above)
Comment (wildjames) [reply packages/cpsuLambda/src/schema/format_1/transformer.ts:15]: Good catch - thanks

File: packages/cpsuLambda/src/schema/format_1/transformer.ts (hunk 2 above)
Comment (wildjames) [reply packages/cpsuLambda/src/schema/format_1/transformer.ts:89]: The ticked specifies only to ignore "Expired" or "NotDispensed", but I'll check with Pete when he's back

File: packages/cpsuLambda/src/schema/format_1/transformer.ts (hunk 3 above)
Comment (wildjames) [reply packages/cpsuLambda/src/schema/format_1/transformer.ts]: Altered the logging message to be a bit more specific

File: packages/cpsuLambda/src/schema/format_1/transformer.ts [hunk 5]
```diff
@@ -78,10 +79,23 @@ export function generateTemplate(requestBody: requestType): string {
...
 
+  if (prescriptionItem.status === "DispensingComplete") {
+    const forbiddenStatuses = ["Expired", "NotDispensed"]
+
+    if (prescriptionItem.completedStatus && forbiddenStatuses.includes(prescriptionItem.completedStatus)) {
+      _logger.info("Skipping data store update for DispensingComplete - completedStatus is an ignored value", {
+        itemID: prescriptionItem.itemID,
+        completedStatus: prescriptionItem.completedStatus
+      })
+      return undefined
```
Comment (MatthewPopat-NHS) [inline packages/cpsuLambda/src/schema/format_1/transformer.ts:95]: we should be using null not undefined

File: packages/cpsuLambda/src/schema/format_1/transformer.ts [hunk 6]
```diff
@@ -78,10 +79,23 @@ export function generateTemplate(requestBody: requestType): string {
...
   prescriptionItem: itemType,
-  prescriptionDetails: requestType
-): Result<BundleEntry<Task>, string> {
+  prescriptionDetails: requestType,
+  _logger: Logger
+): Result<BundleEntry<Task>, string> | undefined {
   const entry = JSON.parse(template) as BundleEntry<Task>
 
+  if (prescriptionItem.status === "DispensingComplete") {
+    const forbiddenStatuses = ["Expired", "NotDispensed"]
```
Comment (MatthewPopat-NHS) [inline packages/cpsuLambda/src/schema/format_1/transformer.ts:88]: This should include cancelled

File: packages/cpsuLambda/src/schema/format_1/transformer.ts (hunk 5 above)
Comment (wildjames) [reply packages/cpsuLambda/src/schema/format_1/transformer.ts]: I've returned an `Err`, and filtered them out when the populated templates are collected.

File: packages/cpsuLambda/src/schema/format_1/transformer.ts [hunk 7]
```diff
@@ -5,20 +5,30 @@ import {
...
+  collectMaybe,
+  Just,
+  Maybe,
+  Nothing,
+  Ok
+} from "pratica"
 import {v4 as uuidv4} from "uuid"
 import {Transformer} from "../../handler"
-import {wrap_with_status} from "../../utils"
+import "../../utils"
```
Comment (wildjames) [inline packages/cpsuLambda/src/schema/format_1/transformer.ts:17]: note: this is so we can use the String.window functionality in the UUID generator

**Review by MatthewPopat-NHS — APPROVED** [review]
